### PR TITLE
Add a FetchHttpClient

### LINKED ISSUE #9444: [SignalR] Consider adding an HttpClient implementation that uses fetch instead of XHR
We should also consider if/how to polyfill on browsers that don't support it.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/aspnet/AspNetCore/issues/9444

Taking over PR #12599, I'll try to keep the two original commits when merging.

### CODE REVIEW

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts [hunk 1]
```diff
@@ -49,12 +49,12 @@ function getConnectionBuilder(transportType?: HttpTransportType, url?: string, o
 }
 
 describe("hubConnection", () => {
-    eachTransportAndProtocol((transportType, protocol) => {
+    eachTransportAndProtocolAndHttpClient((transportType, protocol, httpClient) => {
```
Comment (halter73) [inline src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: How many new test cases does this add? How much longer does this make the tests run?

File: src/SignalR/clients/ts/signalr/src/FetchHttpClient.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        } finally {
+            if (timeoutId) {
+                clearTimeout(timeoutId);
+            }
+        }
+
+        if (!response.ok) {
+            return new HttpError(response.statusText, response.status);
+        } else {
+            if (request.abortSignal) {
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/FetchHttpClient.ts]: Technically you can call clearTimeout with null too.

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: ~100 tests, took ~13 seconds

Comment (halter73) [reply src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: I'll allow it 😉. We can always do a cleanup later where we remove extranious variations.

File: src/SignalR/clients/ts/signalr/src/FetchHttpClient.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                clearTimeout(timeoutId);
+            }
+            if (request.abortSignal) {
+                request.abortSignal.onabort = null;
+            }
+        }
+
+        if (!response.ok) {
+            throw new HttpError(response.statusText, response.status);
+        } else {
```
Comment (davidfowl) [inline src/SignalR/clients/ts/signalr/src/FetchHttpClient.ts]: nit: remove else.

**Comment by BrennanConroy** [comment]
Bueller

**Review by davidfowl — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

**Pull request merged**